### ISSUE TITLE
Tarantool vitality while recovering box.cfg

### ISSUE THREAD

**Issue by rosik** [tarantool/tarantool#4646]
Tarantool instance can be accessed by netbox while it recovers snapshot:

```text
2019-11-25 13:37:29.727 [9576] iproto/101/main I> binary: bound to 0.0.0.0:3301
2019-11-25 13:37:29.728 [9576] main/102/interactive I> recovery start
2019-11-25 13:37:29.728 [9576] main/102/interactive I> recovering from `./00000000000191815704.snap'
2019-11-25 13:37:29.729 [9576] main/102/interactive I> cluster uuid c09773c1-fb89-450b-b4b9-0519458bdfa7
2019-11-25 13:37:29.828 [9576] main/102/interactive I> assigned id 1 to replica 2d4a4420-9b5c-4dd8-baf9-d3af8327a059
2019-11-25 13:37:29.922 [9576] main/102/interactive I> 0.1M rows processed
2019-11-25 13:37:30.006 [9576] main/102/interactive I> 0.2M rows processed
2019-11-25 13:37:30.098 [9576] main/102/interactive I> 0.3M rows processed
2019-11-25 13:37:30.187 [9576] main/102/interactive I> 0.4M rows processed
2019-11-25 13:37:30.270 [9576] main/102/interactive I> 0.5M rows processed
```

During rows processing I can do `tarantoolctl connect localhost:3301` and inspect instance status:

```console
localhost:3301> box.info.status
---
- loading
...
```

But when processing snapshot finishes and tarantool starts building secondary indexes it **becomes unresponsive**. I can't see why behavior is different in these two cases. Maybe it's a bug?

**Comment by kostja** [comment]
We do a parallel sort and there are no yields during it. Yields could be added, but it is not trivial.

**Comment by parihaaraka** [comment]
A caller needs to know if the whole initialization is finished (not box.cfg only). It would be nice to have a field within an authentication response to be settled manually after all the modules loaded.

**Comment by kostja** [comment]
Responding to user queries during snapshot recovery was implemented in response to a feature request. So it's fun that one user declares a bug what another user found to be  a useful feature. In any case, since it was a requested feature, it means someone out there may depend on it so it's a bad idea to simply break it.

But even if you just decide to break it and ban box.cfg.listen queries during snapshot recovery, it won't help. We need to listen to clients during WAL recovery, since otherwise we'll never be able to leave orphan mode. 
So a proper fix has to distinguish peers and clients when accepting connections. Connections from peers, the monitoring scripts and administrative applications, such as cartridge, should be let through. Connections from clients, such as net.box and vshard router data  connections, should be put on hold or be limited in the number of file descriptors they use.

**Comment by rosik** [comment]
I don't get you. What feature are you talking about?

In my opinion *unresponsiveness* during building indexes is the bug. I'll never believe it was made intentionally.

Listening during recovery is certainly good, and answering eval `2 + 2` should be valid even before recovery starts. Database + application server, do you remember? But listening `~=` answering select requests with inconsistent data.

**Comment by kostja** [comment]
I agree we should respond when building indexes, but I'm not sure it's possible to do as long as the implementation uses openmp. I can not find the feature request which adds query handling to the recovery step.
